lib/lf/geometry/tria_o2.cc

```diff
@@ -0,0 +1,225 @@
+/**
+ * @file
+ * @brief Implementation of second-order parametric triangles
+ * @author Anian Ruoss
+ * @date   2018-12-05 22:37:17
+ * @copyright MIT License
+ */
+
+#include "tria_o2.h"
+#include "point.h"
+
+namespace lf::geometry {
+
+TriaO2::TriaO2(Eigen::Matrix<double, Eigen::Dynamic, 6> coords)
+    : coords_(std::move(coords)),
+      alpha_(coords_.rows()),
+      beta_(coords_.rows(), 2),
+      gamma_(coords_.rows(), 2),
+      delta_(coords_.rows()),
+      gamma_x_2_(coords_.rows(), 2) {
+  /*
+   *  2                                C
+   *  | \                              | \
+   *  5   4              ->            F   E
+   *  |     \                          |     \
+   *  0 - 3 - 1                        A - D - B
+   */
+  const Eigen::VectorXd& A = coords_.col(0);
+  const Eigen::VectorXd& B = coords_.col(1);
+  const Eigen::VectorXd& C = coords_.col(2);
+  const Eigen::VectorXd& D = coords_.col(3);
+  const Eigen::VectorXd& E = coords_.col(4);
+  const Eigen::VectorXd& F = coords_.col(5);
+
+  alpha_ << A;
+  beta_ << 4. * D - 3. * A - B, 4. * F - 3. * A - C;
+  gamma_ << 2. * (A + B) - 4. * D, 2. * (A + C) - 4. * F;
+  delta_ << 4. * (A + E - D - F);
+
+  // coefficients for Jacobian
+  gamma_x_2_ << 2. * gamma_;
+}
+
+Eigen::MatrixXd TriaO2::Global(const Eigen::MatrixXd& local) const {
+  LF_VERIFY_MSG((0. <= local.array()).all() && (local.array() <= 1.).all(),
+                "local coordinates out of bounds for reference element");
+
+  return ((beta_ * local) + (gamma_ * local.array().square().matrix()) +
+          (delta_ * local.row(0).cwiseProduct(local.row(1))))
+             .colwise() +
+         alpha_;
+}
+
+Eigen::MatrixXd TriaO2::Jacobian(const Eigen::MatrixXd& local) const {
+  LF_VERIFY_MSG((0. <= local.array()).all() && (local.array() <= 1.).all(),
+                "local coordinates out of bounds for reference element");
+
+  Eigen::MatrixXd tmp(gamma_.rows(), 2 * local.cols());
+
+  for (int i = 0; i < local.cols(); ++i) {
+    tmp.block(0, 2 * i, tmp.rows(), 2) =
+        gamma_x_2_.array().rowwise() * local.col(i).transpose().array();
```

Comment:
I don't quite understand what you are trying to accomplish by using `Eigen::Map` with `tmp`. This is what I'm trying to do:

```
gamma_x_2_ = [[ a, b ],
              [ c, d ]]

local = [[x1, x3, x5],
         [x2, x4, x6]]

=> [[x1*a, x2*b, x3*a, x4*b, x5*a, x6*b],
    [x1*c, x2*d, x3*c, x4*d, x5*c, x6*d]]
```

I tried mapping `gamma_x_2_` to an `Eigen::VectorXd` and the mutliplying with `local.replicate(gamma_.rows(), 1).colwise()` but this gives me a `4xN` instead of `2x2N` matrix which cannot be reshaped correctly due to the storage order.

Reply:
You could e.g. write
```
tmp.row(0) = a*local.row(0);
tmp.row(1) = c*local.row(0);
tmp.row(2) = b*local.row(1);
tmp.row(3) = d*local.row(1);
```
Or am I missing something?

Reply:
Yes, I'm sorry my example was not very clear. For `DimGlobal() = 3` gamma_x_2_ is a `3x2` matrix...

Reply:
@craffael bump